### PR TITLE
Change submit to create and update 

### PR DESCRIPTION
**Pull request by elsaperelli**
# Summary
Because of developments in a decent CRMI call, we want to move away from implementing a $submit operation on the backend and simply allow the measure repository service to support PUT/POST of a new resource (Library or Measure).

## New behavior
The measure repository service now supports PUT/POST of a new resource, where the former uses the ID provided by the requester, and the latter generates a new server ID. This functionality is implemented through `create` and `update` which are already defined in `src/types/service.ts`. This functionality is similar to that in [deqm-test-server](https://github.com/projecttacoma/deqm-test-server).

## Code changes
- `backend/README.md` - Removed $submit info, added create and update info.
- `backend/src/config/capabilityStatementResources.json` - removed submit operation.
- `backend/src/config/serverConfig.ts` - removed submit operation.
- `dbOperations.ts` - added updateResource function and changed `logger.info` to `console.log`. Is this okay? I only changed it because I was getting a winston warning with `logger.info`.
- `LibraryService.ts` / `MeasureService.ts` - added create and update functions.
- `LibraryService.test.ts` / `MeasureService.test.ts` - added unit tests for create and update.
- `utils.ts` - export `createTestResource` function.

# Testing guidance
- If you aren't sure if you have the ColorectalCancerScreeningsFHIR bundle in your database, I would reset the database:
     - `cd backend`
     - `npm run db:reset`
     - `npm run db:loadBundle "ecqm-content-r4-2021/bundles/measure/ColorectalCancerScreeningsFHIR/ColorectalCancerScreeningsFHIR-bundle.json"`
- In the root directory, run: `npm run check:all`
     - Make sure the unit tests pass.
- In the root directory, run: `npm run start:backend`
- Send some PUT and POST requests to the new endpoints in Insomnia. Attached below are some example requests you can import into Insomnia.

[update-and-create-testing.json.zip](https://github.com/projecttacoma/measure-repository/files/11043801/update-and-create-testing.json.zip)

### CODE REVIEW

File: backend/test/services/LibraryService.test.ts [hunk 1]
```diff
@@ -180,6 +180,74 @@ describe('LibraryService', () => {
...
+        .send({ resourceType: 'Library', id: 'exampleId', status: 'active' })
+        .set('content-type', 'application/json+fhir')
+        .expect(200);
+      await supertest(server.app)
+        .get('/4_0_1/Library/exampleId')
+        .expect(200)
+        .then(response => {
+          expect(response.body.id).toEqual('exampleId');
+        });
+    });
```
Comment (natjoe4) [inline backend/test/services/LibraryService.test.ts:213]: I think this second supertest call isn't necessary here. Just ensuring that we get back a 200 response seems sufficient to me for this test.

File: backend/test/services/LibraryService.test.ts [hunk 2]
```diff
@@ -180,6 +180,74 @@ describe('LibraryService', () => {
...
+        });
+      await supertest(server.app)
+        .get('/4_0_1/Library/newId')
+        .expect(200)
+        .then(response => {
+          expect(response.body.resourceType).toEqual('Library');
+          expect(response.body.id).toEqual('newId');
+          expect(response.body.status).toEqual('draft');
+        });
+    });
```
Comment (natjoe4) [inline backend/test/services/LibraryService.test.ts:232]: Same here, this seems like getting a bit close to integration testing. I think if we can test that the headers and status code are what we expect, we can trust mongo and asymmetrik to work as advertised in terms of actually storing the resource and making it accessible.

File: backend/test/services/LibraryService.test.ts [hunk 3]
```diff
@@ -180,6 +180,74 @@ describe('LibraryService', () => {
...
+  });
+
+  describe('update', () => {
+    it('returns 200 when provided correct headers and a FHIR Library whose id is in the database', async () => {
+      await createTestResource(
+        { resourceType: 'Library', type: { coding: [{ code: 'logic-library' }] }, id: 'exampleId', status: 'draft' },
+        'Library'
+      );
```
Comment (natjoe4) [inline backend/test/services/LibraryService.test.ts:201]: I would put this resource creation into a `beforeAll` call in this describe block. Not sure there's anything wrong with the way you've done it here, but that's been our standard for creating test resources in the past.

File: backend/test/services/MeasureService.test.ts [hunk 4]
```diff
@@ -172,6 +172,71 @@ describe('MeasureService', () => {
...
+        .expect(201)
+        .then(response => {
+          expect(response.headers.location).toBeDefined();
+        });
+    });
+  });
+
+  describe('update', () => {
+    it('returns 200 when provided correct headers and a FHIR Measure whose id is in the database', async () => {
+      await createTestResource({ resourceType: 'Measure', id: 'exampleId', status: 'draft' }, 'Measure');
```
Comment (natjoe4) [inline backend/test/services/MeasureService.test.ts:190]: Same here re `beforeAll`

File: backend/test/services/LibraryService.test.ts [hunk 5]
```diff
@@ -180,6 +180,74 @@ describe('LibraryService', () => {
...
+      await supertest(server.app)
+        .put('/4_0_1/Library/exampleId')
+        .send({ resourceType: 'Library', id: 'exampleId', status: 'active' })
+        .set('content-type', 'application/json+fhir')
+        .expect(200);
```
Comment (natjoe4) [inline backend/test/services/LibraryService.test.ts:206]: I would add a location header check here

File: backend/test/services/MeasureService.test.ts [hunk 6]
```diff
@@ -172,6 +172,71 @@ describe('MeasureService', () => {
...
+      await supertest(server.app)
+        .put('/4_0_1/Measure/exampleId')
+        .send({ resourceType: 'Measure', id: 'exampleId', status: 'active' })
+        .set('content-type', 'application/json+fhir')
+        .expect(200);
```
Comment (natjoe4) [inline backend/test/services/MeasureService.test.ts:195]: A location header check here too

**Review by natjoe4 — CHANGES_REQUESTED** [review]
Functionality looks great. Just a few inline comments. Nice work!

**Review by natjoe4 — APPROVED** [review]
Epic work

File: backend/src/config/capabilityStatementResources.json [hunk 7]
```diff
@@ -123,16 +123,6 @@
           "name": "package",
           "definition": "http://hl7.org/fhir/us/cqfmeasures/OperationDefinition/Library-package"
         },
-        {
```
Comment (mgramigna) [inline backend/src/config/capabilityStatementResources.json:126]: We might want to add support for FHIR `create`/`update` to the capability statement now

File: backend/src/services/MeasureService.ts [hunk 8]
```diff
@@ -50,6 +50,39 @@ export class MeasureService implements Service<fhir4.Measure> {
...
+   * result of sending a POST request to {BASE_URL}/4_0_1/Measure
+   * creates a new Measure resource, generates an id for it, and adds it to the database
+   */
+  async create(_: RequestArgs, { req }: RequestCtx) {
+    logger.info('POST /Measure');
+    const contentType: string | undefined = req.headers['content-type'];
+    checkContentTypeHeader(contentType);
+    const resource = req.body;
+    checkExpectedResourceType(resource.resourceType, 'Measure');
+    // create new id
```
Comment (mgramigna) [inline backend/src/services/MeasureService.ts:63]:
```suggestion
```

**Review by mgramigna — COMMENTED** [review]
Looks good, small comments

File: backend/src/config/capabilityStatementResources.json (hunk 7 above)
Comment (mgramigna) [reply backend/src/config/capabilityStatementResources.json:126]: I think they're already there? Unless I'm missing somethign

**Review by mgramigna — APPROVED** [review]

**Pull request merged**